introduces request and response format and deprecates default_format

theocodes opened the pull request:
Recap..

As per my issue #111  and suggestion from Luca, this PR introduces `default_request_format` and `response_request_format` and deprecates `default_format`. 

The premise behind it is:

`default_request_format` behaves the same as `default_format`. However, now it's clear what default format we're setting. request..

`default_response_format` will set it's value to all responses. It can only be overwritten by directly setting `self.format = :whatever` in the action itself. If a value is not set, it will then behave as it did before (try to guess response format based on request format etc...)

`default_format` has been deprecated and should be removed eventually.

Please have a look and let me know what you think. :-)

Notes:

* There's a PR on lotus/lotus that addresses the same thing.
* I've versioned the new methods to 0.5.0 as I assume that's when it will be introduced, but that needs to be confirmed. thoughts?

Closes #111

jodosha commented on test/configuration_test.rb:
```diff
@@ -324,21 +346,21 @@ def hash
+      @config.action_module.must_equal          @configuration.action_module
+      @config.modules.must_equal                @configuration.modules
+      @config.send(:formats).must_equal         @configuration.send(:formats)
+      @config.default_request_format.must_equal @configuration.default_request_format
```
Can you please test `default_response_format` too? :smile:

jodosha commented on test/configuration_test.rb:
```diff
@@ -347,7 +369,7 @@ def hash
       @config.prepare          { include Comparable }
       @config.format another: 'another/format'
-      @config.default_format :json
+      @config.default_request_format :json
```
Same as: https://github.com/lotus/controller/pull/125/files#r39599214

jodosha commented on test/configuration_test.rb:
```diff
@@ -356,7 +378,7 @@ def hash
       @config.modules.size.must_equal                2
       @config.format_for('another/format').must_equal :another
-      @config.default_format.must_equal               :json
+      @config.default_request_format.must_equal               :json
```
Same as: https://github.com/lotus/controller/pull/125/files#r39599214

jodosha commented on test/configuration_test.rb:
```diff
@@ -365,7 +387,7 @@ def hash
       @configuration.modules.size.must_equal       1
       @configuration.format_for('another/format').must_be_nil
-      @configuration.default_format.must_equal     :html
+      @configuration.default_request_format.must_equal :html
```
Same as: https://github.com/lotus/controller/pull/125/files#r39599214

jodosha commented on test/configuration_test.rb:
```diff
@@ -378,7 +400,7 @@ def hash
       @configuration.modules          { include Kernel }
       @configuration.format another: 'another/format'
-      @configuration.default_format :another
+      @configuration.default_request_format :another
```
Same as: https://github.com/lotus/controller/pull/125/files#r39599214

jodosha commented on test/configuration_test.rb:
```diff
@@ -391,7 +413,7 @@ def hash
       @configuration.modules.must_equal([])
       @configuration.send(:formats).must_equal(Lotus::Controller::Configuration::DEFAULT_FORMATS)
-      @configuration.default_format.must_be_nil
+      @configuration.default_request_format.must_be_nil
```
Same as: https://github.com/lotus/controller/pull/125/files#r39599214

jodosha commented on test/integration/mime_type_test.rb:
```diff
@@ -111,6 +123,12 @@ def call(params)
+    response.headers['Content-Type'].must_equal 'application/json; charset=utf-8'
+    response.body.must_equal                    'html'
+  end
+
```
I think we should also assert that `self.format=` will be honored.

```ruby
class OverrideDefaultResponse
  include Lotus::Action

  configuration.default_response_format :json

  def call(params)
    self.body = :xml
  end
end

it 'allows to override default_response_format' do
  response = @app.get('/overwritten_format')
  response.headers['Content-Type'].must_equal 'application/xml; charset=utf-8'
end
```

theocodes replied:
Agreed!  Thanks  :+1:

jodosha commented:
> I've versioned the new methods to 0.5.0 as I assume that's when it will be introduced, but that needs to be confirmed. thoughts?

This is fine right now, but please use `@since x.x.x` instead. Before a release we grep for these YARD tags and replace with the right value :smile:

jodosha commented:
@theocodes This looks good, I left a few comments, can you please address these issues?

jodosha commented on test/configuration_test.rb:
```diff
@@ -187,25 +187,47 @@ def hash
     end
   end
 
-  describe '#default_format' do
```
@theocodes We need both the tests for `#default_format` and `#default_request_format`. The first is still there, and we want to make sure that it still works. :wink:

jodosha commented on lib/lotus/controller/configuration.rb:
```diff
@@ -418,35 +418,92 @@ def format(hash)
+      #
+      # @deprecated Use {#default_request_format} instead.
       def default_format(format = nil)
+        Lotus::Utils::Deprecation.new('default_format is deprecated, please use default_request_format')
```
@theocodes We probably need a test case for this deprecation. You can use Minitest `capture_io`.

```ruby
_, err = capture_io do
  Lotus::Controller.configure do
    default_format :png
  end
end

err.chomp.must_equal "..."
```

theocodes commented:
@jodosha I believe I have addressed all the issues.  :)

Note: There's a lot of noise about that deprecated method when you run the test suite now.. Is there way to mute it for tests only?

jodosha commented:
@theocodes This looks good, thank you :+1:
